### DISCUSSION TITLE
Interaction and ordering of Kademlia, Identfiy and AutoNAT behaviors

### DISCUSSION THREAD

**Discussion by b-zee** [libp2p/rust-libp2p#3906] [Q&A] [answered]
Assume these three behaviors: Kademlia, Identify and AutoNAT. Adding peers to the Kademlia routing table happens based on the `Info` we exchange (out-of-band) with Identify. However, we only want to add the proper address of a peer. For a peer to first determine its public address, it uses AutoNAT.

The addresses that get broadcast with `Info` include our local addresses, which can be ignored by our peers. But only after we have established confidently our external address the `Info` will also include that address. We can [`push`](https://docs.rs/libp2p/latest/libp2p/identify/struct.Behaviour.html#method.push) the `Info` again when we see a change in NAT status. However, this [`status`](https://docs.rs/libp2p/latest/libp2p/autonat/enum.NatStatus.html) doesn't list the confidence. And there is no swarm event that I know of that we can use to see if external addresses were added or the scores have changed.

Am I going about this in the wrong way? In essence, what I am trying to prevent is a node behind NAT trying to join the Kademlia network. I am assuming that will be detrimental to the health of that network, as the node can not be dialed (assuming we are not yet using relays).

**Comment by thomaseizinger** [comment, marked as answer]
Kademlia will only ever add a node to the routing table if it successfully dialed that node:

https://github.com/libp2p/rust-libp2p/blob/541eba306b251a66f08ab2df0dcd5f753aa1a553/protocols/kad/src/behaviour.rs#L2050-L2057

The addresses we receive from identify will be use to attempt to connect to a node if we don't have a connection yet:

https://github.com/libp2p/rust-libp2p/blob/541eba306b251a66f08ab2df0dcd5f753aa1a553/protocols/identify/src/behaviour.rs#L330-L343

which gets populated by:

https://github.com/libp2p/rust-libp2p/blob/541eba306b251a66f08ab2df0dcd5f753aa1a553/protocols/identify/src/behaviour.rs#L286-L293

Does this help?

**Reply by b-zee** [reply]
Amazing, thanks for the pointers.

> Kademlia will only ever add a node to the routing table if it successfully dialed that node

Does this mean that clients are attempted to be dialed? (In case of using QUIC or TCP with port reuse, that might mean that those clients are actually successfully dialed. (Even assuming NAT, as the hole will be punched because the port is reused.))

**Reply by thomaseizinger** [reply]
> > Kademlia will only ever add a node to the routing table if it successfully dialed that node
> 
> Does this mean that clients are attempted to be dialed?

If you look for usages of `ToSwarm::Dial` in the kademlia protocol, you'll find the following three:

1. We are working on a query and need to send a message to a peer we are not yet connected to:

https://github.com/libp2p/rust-libp2p/blob/3fde13220b29f0878d6f98a3fe50bc9b008baf8a/protocols/kad/src/behaviour.rs#L2402-L2404

2. We evicted an entry from the routing table because a bucket was full, thus we test connectivity to that node:

https://github.com/libp2p/rust-libp2p/blob/3fde13220b29f0878d6f98a3fe50bc9b008baf8a/protocols/kad/src/behaviour.rs#L574-L576
https://github.com/libp2p/rust-libp2p/blob/3fde13220b29f0878d6f98a3fe50bc9b008baf8a/protocols/kad/src/behaviour.rs#L1212-L1216

**Reply by b-zee** [reply]
Thanks again, sorry if that was a bit low effort. I am reading through the code myself now, and getting the hang of tracing these things.

I think that mostly clears it up for now. I will try and see if I can get these basic behaviors interacting with each other.

**Reply by mxinden** [reply]
Thanks @thomaseizinger :+1: 

> I will try and see if I can get these basic behaviors interacting with each other.

I am sorry for the trouble @b-zee you are facing with some of this. Note that the `libp2p-autonat` implementation is relatively new and not much tested in the wild.

**Reply by thomaseizinger** [reply]
> Thanks again, sorry if that was a bit low effort. I am reading through the code myself now, and getting the hang of tracing these things.

No worries at all :)

It is a big codebase, feel free to ask if more questions arise!

**Comment by b-zee** [comment]
@thomaseizinger 

You mentioned that peers only get added when they are dialed by us:
> Kademlia will only ever add a node to the routing table if it successfully dialed that node

Though, this seems to be overwritten of course if we manually call [`add_address`](https://docs.rs/libp2p/latest/libp2p/kad/struct.Kademlia.html#method.add_address). This was something I overlooked in our codebase.

Now, how would we approach leaving this to the Kademlia protocol. Does Kademlia add all nodes that support the `kad` protocol and are dialed by us? Or is this only in case this dial was part of a query? E.g. we called `add_address` manually based on the `identify` protocol to see if it's a peer with a certain user agent string, but we'd still only want to add them if we can successfully dial them.

**Reply by thomaseizinger** [reply]
> Now, how would we approach leaving this to the Kademlia protocol. Does Kademlia add all nodes that support the `kad` protocol and are dialed by us? Or is this only in case this dial was part of a query? E.g. we called `add_address` manually based on the `identify` protocol to see if it's a peer with a certain user agent string, but we'd still only want to add them if we can successfully dial them.

I am working on this being automatic, yes: https://github.com/libp2p/rust-libp2p/pull/3877

Should be done soon (hopefully) :)

**Reply by b-zee** [reply]
Cool, thanks for replying!

Does this mean that a successful `dial` to a peer will add the peer to the routing table (assuming the right protocol etc)? That is, no `add_server` needed?

Is there a workaround for now to achieve this behavior? I don't think we can see whether we are the `Dialer`, so it's only possible to keep track of this ourselves using `ConnectionEstablished` events?

**Reply by thomaseizinger** [reply]
> Cool, thanks for replying!
> 
> Does this mean that a successful `dial` to a peer will add the peer to the routing table (assuming the right protocol etc)? That is, no `add_server` needed?

Yes!

> Is there a workaround for now to achieve this behavior? I don't think we can see whether we are the `Dialer`, so it's only possible to keep track of this ourselves using `ConnectionEstablished` events?

The `ConnectedPoint` exposes dialer/listener :)

**Reply by b-zee** [reply]
Yeah, so it's something we'd track in a list manually. I think that's a solid workaround for now unless you know of something simpler.

Thanks a lot for the help! Great work you guys are doing!
